Remove default charset from 'application/json' Content-Type header

doortts opened the pull request:
Using application/json; charset=UTF-8 in response header is a common misuse. I think it is better to remove `; charset=UTF-8` from default json response Content-Type header to prevent the misconception.

See: https://github.com/labstack/echo/issues/2567

aldas reviewed (changes requested):
You can not just remove/replace `echo.MIMEApplicationJSONCharsetUTF8` as it is part of Echos public API and we are quite strict on API changes.

+ it would be nice if there would be link to JSON spec and remark about charset in doccomment


other than that I think this is probably ok

doortts commented:
@aldas Thank you for the comment. 

- I recovered `echo.MIMEApplicationJSONCharsetUTF8` and added 'Deprecated: ' comment. Just let let me know if this is not what you intended.  https://github.com/labstack/echo/pull/2568/files?diff=unified#diff-aba8ba313c6b989153f1f96e641a8e0e75313485d2e13d64378e70858d5a4669R173
- I added a reference link what you mentioned to the issue documentation.
   - [Related RFC note link](https://www.rfc-editor.org/rfc/rfc8259#:~:text=Note%3A%20%20No%20%22charset%22%20parameter%20is%20defined%20for%20this%20registration.%0A%20%20%20%20%20%20Adding%20one%20really%20has%20no%20effect%20on%20compliant%20recipients.)
   - [Related RFC 8.1 Chapter Link](https://www.rfc-editor.org/rfc/rfc8259#section-8.1:~:text=JSON%20text%20exchanged%20between%20systems%20that%20are%20not%20part%20of%20a%20closed%0A%20%20%20ecosystem%20MUST%20be%20encoded%20using%20UTF%2D8)

aldas commented on echo.go:
```diff
@@ -169,7 +169,8 @@ const (
 // MIME types
 const (
-	MIMEApplicationJSON                  = "application/json"
+	MIMEApplicationJSON = "application/json"
```
I did not express correctly. What I meant that it would be helpful to have these rfc link also in doccomments. Something like that 

```go
  // JavaScript Object Notation (JSON) https://www.rfc-editor.org/rfc/rfc8259
	MIMEApplicationJSON = "application/json"
	// Deprecated: use MIMEApplicationJSON instead. JSON should be encoded using UTF-8 by RFC https://www.rfc-editor.org/rfc/rfc8259#section-8.1
	MIMEApplicationJSONCharsetUTF8       = MIMEApplicationJSON + "; " + charsetUTF8
```

doortts replied:
Ah, I see. I've modified the code according to your suggestion.

How about this?
```go
  // MIMEApplicationJSON JavaScript Object Notation (JSON) https://www.rfc-editor.org/rfc/rfc8259
	MIMEApplicationJSON = "application/json"
	// Deprecated: Please use MIMEApplicationJSON instead. JSON should be encoded using UTF-8.
	// No "charset" parameter is defined for this registration.
	// Adding one really has no effect on compliant recipients.
	// See RFC 8259, section 8.1. https://datatracker.ietf.org/doc/html/rfc8259#section-8.1
	MIMEApplicationJSONCharsetUTF8       = MIMEApplicationJSON + "; " + charsetUTF8
```
https://github.com/labstack/echo/pull/2568/files#diff-aba8ba313c6b989153f1f96e641a8e0e75313485d2e13d64378e70858d5a4669R172-R177

doortts replied:
@aldas would you mind review my updated PR?

aldas commented:
done, Thanks @doortts

washington-guedes commented:
There are clients outside of what was called "compliant recipient" that got broken, like flutter v2.1 in my case. Is it possible to have a middleware that forces the header Content-Type back to MIMEApplicationJSONCharsetUTF8? @aldas @doortts

aldas commented:
or you could create custom json serializer than sets that header before writing the json

```go
package main

import (
	"encoding/json"
	"errors"
	"github.com/labstack/echo/v4"
	"log"
	"net/http"
)

type utf8JSONSerializer struct {
	echo.DefaultJSONSerializer
}

func (d utf8JSONSerializer) Serialize(c echo.Context, i interface{}, indent string) error {
	enc := json.NewEncoder(c.Response())
	if indent != "" {
		enc.SetIndent("", indent)
	}
	c.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSONCharsetUTF8)
	return enc.Encode(i)
}

func main() {
	e := echo.New()
	e.JSONSerializer = new(utf8JSONSerializer)

	e.GET("/", func(c echo.Context) error {
		return c.JSON(http.StatusOK, "HI")
	})

	if err := e.Start(":8080"); err != nil && !errors.Is(err, http.ErrServerClosed) {
		log.Fatal(err)
	}
}
```
